plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergTableProperties.java

```diff
@@ -252,6 +297,48 @@ private static void validateFormatVersion(int version)
         }
     }
 
+    public static void validateCompression(IcebergFileFormat fileFormat, Optional<HiveCompressionCodec> compressionCodec, OptionalInt compressionLevel)
+    {
+        if (compressionCodec.isPresent()) {
+            if (!isCompressionCodecSupportedForFormat(fileFormat, compressionCodec.get())) {
+                throw new TrinoException(NOT_SUPPORTED, format("Compression codec LZ4 not supported for %s", fileFormat.humanName()));
+            }
+        }
+        if (compressionLevel.isPresent()) {
+            if (compressionCodec.isEmpty()) {
+                throw new TrinoException(INVALID_TABLE_PROPERTY, "write_compression must be set when compression_level is set");
+            }
+            else {
+                if (!(VALID_ICEBERG_FILE_FORMATS_FOR_COMPRESSION_LEVEL_PROPERTY.contains(fileFormat)
+                        && VALID_ICEBERG_COMPRESSION_CODECS_FOR_COMPRESSION_LEVEL_PROPERTY.contains(compressionCodec.get()))) {
+                    throw new TrinoException(NOT_SUPPORTED,
+                            format("File format %s with compression codec %s and compression level %d not supported", fileFormat, compressionCodec.get(), compressionLevel.getAsInt()));
+                }
+
+                if (compressionCodec.get().equals(HiveCompressionCodec.GZIP)) {
+                    if (!((compressionLevel.getAsInt() >= GZIP_MIN_COMPRESSION_LEVEL) && (compressionLevel.getAsInt() <= GZIP_MAX_COMPRESSION_LEVEL))) {
+                        throw new TrinoException(NOT_SUPPORTED,
+                                format("File format %s with compression codec %s and compression level %d not supported", fileFormat, compressionCodec, compressionLevel.getAsInt()));
+                    }
+                }
+                else if (compressionCodec.get().equals(HiveCompressionCodec.ZSTD)) {
+                    if (!((compressionLevel.getAsInt() >= ZSTD_MIN_COMPRESSION_LEVEL) && (compressionLevel.getAsInt() <= ZSTD_MAX_COMPRESSION_LEVEL))) {
+                        throw new TrinoException(NOT_SUPPORTED,
+                                format("File format %s with compression codec %s and compression level %d not supported", fileFormat, compressionCodec, compressionLevel.getAsInt()));
+                    }
+                }
+                else {
+                    throw new IllegalStateException("Unsupported compression codec: " + compressionCodec);
+                }
+            }
+        }
+    }
+
+    public static boolean isCompressionCodecSupportedForFormat(IcebergFileFormat fileFormat, HiveCompressionCodec codec)
+    {
+        return !((fileFormat.equals(AVRO) || fileFormat.equals(PARQUET)) && codec.equals(HiveCompressionCodec.LZ4));
```

Comment:
Let's rewrite this as a switch over the file formats and compression codecs so that it's easier to read and understand the valid combinations.